File: python/crypto_utils.py
#!/usr/bin/env python3
"""
crypto_utils.py — Cryptographic primitives for secure-p2p

Stage 2: Diffie-Hellman key exchange implemented from scratch.
Stage 3: AES-256-CBC encryption and decryption.

Every mathematical operation is commented so you can explain it in an interview.
"""

import os
import hashlib
import struct


# ─── Diffie-Hellman Parameters (RFC 3526, Group 14 — 2048-bit MODP) ──────────
#
# These two numbers are PUBLIC — hardcoded and identical on both sides.
# Using a standardised prime from RFC 3526 is intentional: it has been
# audited by the cryptography community and is known to be safe.
#
# p: a 2048-bit safe prime. "Safe" means (p-1)/2 is also prime, which
#    prevents certain subgroup attacks against DH.
#
# g: the generator, 2. Every possible shared secret is a power of g mod p.

DH_PRIME_P = int(
    "FFFFFFFF FFFFFFFF C90FDAA2 2168C234 C4C6628B 80DC1CD1"
    "29024E08 8A67CC74 020BBEA6 3B139B22 514A0879 8E3404DD"
    "EF9519B3 CD3A431B 302B0A6D F25F1437 4FE1356D 6D51C245"
    "E485B576 625E7EC6 F44C42E9 A637ED6B 0BFF5CB6 F406B7ED"
    "EE386BFB 5A899FA5 AE9F2411 7C4B1FE6 49286651 ECE45B3D"
    "C2007CB8 A163BF05 98DA4836 1C55D39A 69163FA8 FD24CF5F"
    "83655D23 DCA3AD96 1C62F356 208552BB 9ED52907 7096966D"
    "670C354E 4ABC9804 F1746C08 CA18217C 32905E46 2E36CE3B"
    "E39E772C 180E8603 9B2783A2 EC07A28F B5C55DF0 6F4C52C9"
    "DE2BCBF6 95581718 3995497C EA956AE5 15D22618 98FA0510"
    "15728E5A 8AACAA68 FFFFFFFF FFFFFFFF".replace(" ", ""),
    16
)

DH_GENERATOR_G = 2

# AES block size is always 16 bytes — this is fixed by the AES standard
AES_BLOCK_SIZE = 16


# ─── Diffie-Hellman Key Exchange ──────────────────────────────────────────────

class DiffieHellman:
    """
    Manual implementation of Diffie-Hellman key exchange over a 2048-bit
    safe prime group (RFC 3526, Group 14).
    """

    def __init__(self):
        # Generate a cryptographically random private key.
        # os.urandom(256) gives 256 bytes = 2048 bits of OS-level randomness.
        self._private_key: int = int.from_bytes(os.urandom(256), byteorder="big")
        # Clamp to [2, p-2] to avoid degenerate values
        self._private_key = (self._private_key % (DH_PRIME_P - 2)) + 2

    def get_public_key(self) -> int:
        """
        Compute our public key: g^private_key mod p
        Python's pow(g, e, m) uses fast modular exponentiation — O(log n).
        """
        return pow(DH_GENERATOR_G, self._private_key, DH_PRIME_P)

    def compute_shared_secret(self, their_public_key: int) -> int:
        """
        Compute shared secret: their_public_key^our_private_key mod p
        Both sides arrive at g^(a*b) mod p independently.
        """
        self._validate_public_key(their_public_key)
        return pow(their_public_key, self._private_key, DH_PRIME_P)

    def derive_aes_key(self, shared_secret: int) -> bytes:
        """
        Derive a 256-bit AES key from the raw DH shared secret via SHA-256.
        SHA-256 removes mathematical structure and produces uniform 32 bytes.
        """
        secret_bytes = shared_secret.to_bytes(256, byteorder="big")
        return hashlib.sha256(secret_bytes).digest()

    @staticmethod
    def _validate_public_key(public_key: int) -> None:
        """
        Validate received public key is in range [2, p-2] and in the
        correct subgroup. Prevents small-subgroup confinement attacks.
        """
        if not (2 <= public_key <= DH_PRIME_P - 2):
            raise ValueError("Received public key is out of valid range [2, p-2].")
        order = (DH_PRIME_P - 1) // 2
        if pow(public_key, order, DH_PRIME_P) != 1:
            raise ValueError("Received public key failed subgroup validation.")


# ─── Key serialisation helpers ────────────────────────────────────────────────

def int_to_bytes(n: int) -> bytes:
    """Serialise a DH public key (up to 2048-bit) to 256 bytes, big-endian."""
    return n.to_bytes(256, byteorder="big")


def bytes_to_int(b: bytes) -> int:
    """Deserialise a 256-byte big-endian encoding back to an integer."""
    return int.from_bytes(b, byteorder="big")


# ─── PKCS#7 Padding ───────────────────────────────────────────────────────────

def pkcs7_pad(data: bytes) -> bytes:
    """
    Apply PKCS#7 padding to make data a multiple of AES_BLOCK_SIZE (16 bytes).

    How it works:
      - Calculate how many bytes are needed to reach the next 16-byte boundary
      - Append that many bytes, each with the numeric value of the pad length
      - If data is already a multiple of 16, append a full 16-byte padding block
        (so the receiver can always unambiguously strip padding)

    Example:
      data = b"HELLO" (5 bytes)
      pad_len = 16 - (5 % 16) = 11
      padded  = b"HELLO" + bytes([11] * 11)
    """
    pad_len = AES_BLOCK_SIZE - (len(data) % AES_BLOCK_SIZE)
    return data + bytes([pad_len] * pad_len)


def pkcs7_unpad(data: bytes) -> bytes:
    """
    Remove PKCS#7 padding after decryption.

    The last byte tells us how many padding bytes were appended.
    We validate all padding bytes have the correct value before stripping —
    a padding oracle attack exploits implementations that don't validate this.

    Raises ValueError if padding is malformed (possible tampering).
    """
    if not data:
        raise ValueError("Cannot unpad empty data")

    pad_len = data[-1]  # Last byte = padding length

    # Sanity check: padding length must be 1–16
    if pad_len < 1 or pad_len > AES_BLOCK_SIZE:
        raise ValueError(f"Invalid PKCS#7 padding length: {pad_len}")

    # Verify every padding byte has the correct value
    padding = data[-pad_len:]
    if padding != bytes([pad_len] * pad_len):
        raise ValueError("Invalid PKCS#7 padding — data may have been tampered with")

    return data[:-pad_len]


# ─── AES-256-CBC Encryption / Decryption ─────────────────────────────────────
#
# We use pycryptodome (imported as Crypto) for the raw AES block operations.
# The CBC chaining logic is handled by pycryptodome's CBC mode.
# The IV generation, padding, and framing are our responsibility.

def aes_encrypt(plaintext: bytes, key: bytes) -> bytes:
    """
    Encrypt plaintext using AES-256-CBC.

    Steps:
      1. Generate a random 16-byte IV (never reuse an IV with the same key)
      2. Pad plaintext to a multiple of 16 bytes using PKCS#7
      3. Encrypt with AES-256-CBC using key + IV
      4. Prepend IV to ciphertext (receiver needs it to decrypt)

    Wire format: [16-byte IV][ciphertext]

    Args:
        plaintext: raw bytes to encrypt (any length)
        key:       32-byte AES-256 key (from DH key derivation)

    Returns:
        IV + ciphertext as a single bytes object
    """
    from Crypto.Cipher import AES  # pycryptodome

    if len(key) != 32:
        raise ValueError(f"AES-256 requires a 32-byte key, got {len(key)} bytes")

    # Step 1: Fresh random IV for every encryption — critical for CBC security.
    # Reusing an IV with the same key leaks XOR of the first plaintext blocks.
    iv = os.urandom(AES_BLOCK_SIZE)

    # Step 2: Pad to block boundary
    padded = pkcs7_pad(plaintext)

    # Step 3: Encrypt — pycryptodome handles the CBC XOR chaining internally
    cipher = AES.new(key, AES.MODE_CBC, iv)
    ciphertext = cipher.encrypt(padded)

    # Step 4: Prepend IV so receiver can decrypt
    return iv + ciphertext


def aes_decrypt(ciphertext_with_iv: bytes, key: bytes) -> bytes:
    """
    Decrypt AES-256-CBC ciphertext produced by aes_encrypt().

    Steps:
      1. Split off the first 16 bytes as the IV
      2. Decrypt the remainder with AES-256-CBC using key + IV
      3. Strip PKCS#7 padding to recover original plaintext

    Args:
        ciphertext_with_iv: IV + ciphertext (as returned by aes_encrypt)
        key:                32-byte AES-256 key (must match encryption key)

    Returns:
        Original plaintext bytes
    """
    from Crypto.Cipher import AES  # pycryptodome

    if len(key) != 32:
        raise ValueError(f"AES-256 requires a 32-byte key, got {len(key)} bytes")

    if len(ciphertext_with_iv) < AES_BLOCK_SIZE:
        raise ValueError("Ciphertext too short to contain IV")

    # Step 1: Extract IV from the first 16 bytes
    iv         = ciphertext_with_iv[:AES_BLOCK_SIZE]
    ciphertext = ciphertext_with_iv[AES_BLOCK_SIZE:]

    # Step 2: Decrypt
    cipher    = AES.new(key, AES.MODE_CBC, iv)
    padded    = cipher.decrypt(ciphertext)

    # Step 3: Strip padding
    plaintext = pkcs7_unpad(padded)

    return plaintext


# ─── Handshake orchestration ──────────────────────────────────────────────────

def perform_handshake_sender(sock) -> bytes:
    """
    Sender side of the DH handshake.
    Returns 32-byte AES key derived from the shared secret.
    """
    from peer import send_bytes, recv_bytes

    dh = DiffieHellman()

    our_public_key = dh.get_public_key()
    print("[*] DH: Sending public key to receiver...")
    send_bytes(sock, int_to_bytes(our_public_key))

    their_public_key_bytes = recv_bytes(sock)
    their_public_key = bytes_to_int(their_public_key_bytes)
    print("[*] DH: Received receiver's public key")

    shared_secret = dh.compute_shared_secret(their_public_key)
    print("[*] DH: Shared secret computed")

    aes_key = dh.derive_aes_key(shared_secret)
    print(f"[+] DH: AES-256 key derived — {aes_key.hex()[:16]}... (first 8 bytes shown)")

    return aes_key


def perform_handshake_receiver(sock) -> bytes:
    """
    Receiver side of the DH handshake.
    Returns 32-byte AES key — identical to the sender's derived key.
    """
    from peer import send_bytes, recv_bytes

    dh = DiffieHellman()

    their_public_key_bytes = recv_bytes(sock)
    their_public_key = bytes_to_int(their_public_key_bytes)
    print("[*] DH: Received sender's public key")

    our_public_key = dh.get_public_key()
    print("[*] DH: Sending public key to sender...")
    send_bytes(sock, int_to_bytes(our_public_key))

    shared_secret = dh.compute_shared_secret(their_public_key)
    print("[*] DH: Shared secret computed")

    aes_key = dh.derive_aes_key(shared_secret)
    print(f"[+] DH: AES-256 key derived — {aes_key.hex()[:16]}... (first 8 bytes shown)")

    return aes_key

File: c/crypto_utils.c
#include "crypto_utils.h"
#include "peer.h"

#include <stdio.h>
#include <string.h>
#include <openssl/rand.h>
#include <openssl/evp.h>
#include <openssl/sha.h>

/*
 * RFC 3526 Group 14 — 2048-bit MODP prime (same as Python implementation).
 * Both peers use identical p and g so the handshake is compatible across
 * the C and Python implementations.
 */
static const char *DH_PRIME_HEX =
    "FFFFFFFFFFFFFFFFC90FDAA22168C234C4C6628B80DC1CD1"
    "29024E088A67CC74020BBEA63B139B22514A08798E3404DD"
    "EF9519B3CD3A431B302B0A6DF25F14374FE1356D6D51C245"
    "E485B576625E7EC6F44C42E9A637ED6B0BFF5CB6F406B7ED"
    "EE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3D"
    "C2007CB8A163BF0598DA48361C55D39A69163FA8FD24CF5F"
    "83655D23DCA3AD961C62F356208552BB9ED529077096966D"
    "670C354E4ABC9804F1746C08CA18217C32905E462E36CE3B"
    "E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9"
    "DE2BCBF6955817183995497CEA956AE515D2261898FA0510"
    "15728E5A8AACAA68FFFFFFFFFFFFFFFF";

static const int DH_GENERATOR = 2;


/* ── DH context lifecycle ─────────────────────────────────────────────────── */

int dh_init(dh_ctx_t *ctx) {
    ctx->bn_ctx = BN_CTX_new();
    if (!ctx->bn_ctx) return -1;

    /* Allocate 256 bytes (2048 bits) of cryptographically random data */
    uint8_t rand_buf[DH_KEY_BYTES];
    if (RAND_bytes(rand_buf, DH_KEY_BYTES) != 1) {
        BN_CTX_free(ctx->bn_ctx);
        return -1;
    }

    /* Convert random bytes to a BIGNUM private key */
    ctx->private_key = BN_bin2bn(rand_buf, DH_KEY_BYTES, NULL);
    if (!ctx->private_key) {
        BN_CTX_free(ctx->bn_ctx);
        return -1;
    }

    /*
     * Clamp private key to range [2, p-2].
     * We do this by: private_key = (private_key mod (p-2)) + 2
     * Same logic as the Python implementation.
     */
    BIGNUM *p      = BN_new();
    BIGNUM *p_min2 = BN_new();
    BIGNUM *two    = BN_new();

    BN_hex2bn(&p, DH_PRIME_HEX);
    BN_copy(p_min2, p);
    BN_sub_word(p_min2, 2);
    BN_set_word(two, 2);

    BN_mod(ctx->private_key, ctx->private_key, p_min2, ctx->bn_ctx);
    BN_add(ctx->private_key, ctx->private_key, two);

    BN_free(p);
    BN_free(p_min2);
    BN_free(two);

    /* Clear the random buffer from stack memory */
    memset(rand_buf, 0, sizeof(rand_buf));
    return 0;
}

void dh_free(dh_ctx_t *ctx) {
    if (ctx->private_key) {
        BN_clear_free(ctx->private_key);  /* BN_clear_free zeroes memory first */
        ctx->private_key = NULL;
    }
    if (ctx->bn_ctx) {
        BN_CTX_free(ctx->bn_ctx);
        ctx->bn_ctx = NULL;
    }
}


/* ── DH public key computation ────────────────────────────────────────────── */

int dh_get_public_key(dh_ctx_t *ctx, uint8_t out_buf[DH_KEY_BYTES]) {
    /*
     * public_key = g^private_key mod p
     *
     * BN_mod_exp() uses the Montgomery reduction algorithm for fast
     * modular exponentiation — O(log e) multiplications, same as
     * Python's built-in pow(g, e, p).
     */
    BIGNUM *g   = BN_new();
    BIGNUM *p   = BN_new();
    BIGNUM *pub = BN_new();

    BN_set_word(g, DH_GENERATOR);
    BN_hex2bn(&p, DH_PRIME_HEX);

    if (BN_mod_exp(pub, g, ctx->private_key, p, ctx->bn_ctx) != 1) {
        BN_free(g); BN_free(p); BN_free(pub);
        return -1;
    }

    /* Serialise to fixed DH_KEY_BYTES bytes, big-endian, zero-padded */
    memset(out_buf, 0, DH_KEY_BYTES);
    int num_bytes = BN_num_bytes(pub);
    BN_bn2bin(pub, out_buf + (DH_KEY_BYTES - num_bytes));

    BN_free(g); BN_free(p); BN_free(pub);
    return 0;
}


/* ── DH shared secret computation ────────────────────────────────────────── */

int dh_compute_shared_secret(dh_ctx_t *ctx,
                              const uint8_t their_pub[DH_KEY_BYTES],
                              uint8_t out_buf[DH_KEY_BYTES]) {
    /*
     * shared_secret = their_public_key^our_private_key mod p
     *               = g^(their_private * our_private) mod p
     *
     * Both peers compute the same value because multiplication is
     * commutative — this is the foundation of Diffie-Hellman.
     */
    BIGNUM *their_pub_bn = BN_bin2bn(their_pub, DH_KEY_BYTES, NULL);
    BIGNUM *p            = BN_new();
    BIGNUM *secret       = BN_new();

    BN_hex2bn(&p, DH_PRIME_HEX);

    if (BN_mod_exp(secret, their_pub_bn, ctx->private_key, p, ctx->bn_ctx) != 1) {
        BN_free(their_pub_bn); BN_free(p); BN_free(secret);
        return -1;
    }

    memset(out_buf, 0, DH_KEY_BYTES);
    int num_bytes = BN_num_bytes(secret);
    BN_bn2bin(secret, out_buf + (DH_KEY_BYTES - num_bytes));

    BN_free(their_pub_bn); BN_free(p); BN_free(secret);
    return 0;
}


/* ── AES key derivation ───────────────────────────────────────────────────── */

int dh_derive_aes_key(const uint8_t secret[DH_KEY_BYTES],
                       uint8_t aes_key_out[AES_KEY_BYTES]) {
    /*
     * AES key = SHA-256(shared_secret)
     *
     * SHA-256 maps the 2048-bit shared secret to a uniform 256-bit key.
     * This removes mathematical structure from the raw DH output.
     * Same operation as the Python implementation — keys are compatible.
     */
    SHA256(secret, DH_KEY_BYTES, aes_key_out);
    return 0;
}


/* ── Handshake ────────────────────────────────────────────────────────────── */

int perform_handshake_sender(int sock, uint8_t aes_key_out[AES_KEY_BYTES]) {
    dh_ctx_t dh;
    if (dh_init(&dh) != 0) return -1;

    uint8_t our_pub[DH_KEY_BYTES];
    uint8_t their_pub[DH_KEY_BYTES];
    uint8_t secret[DH_KEY_BYTES];

    /* Step 1: Send our public key */
    dh_get_public_key(&dh, our_pub);
    printf("[*] DH: Sending public key to receiver...\n");
    if (send_bytes(sock, our_pub, DH_KEY_BYTES) != 0) {
        dh_free(&dh); return -1;
    }

    /* Step 2: Receive their public key */
    uint8_t *recv_buf = NULL;
    ssize_t  recv_len = recv_bytes(sock, &recv_buf);
    if (recv_len != DH_KEY_BYTES) {
        free(recv_buf); dh_free(&dh); return -1;
    }
    memcpy(their_pub, recv_buf, DH_KEY_BYTES);
    free(recv_buf);
    printf("[*] DH: Received receiver's public key\n");

    /* Step 3: Compute shared secret */
    dh_compute_shared_secret(&dh, their_pub, secret);
    printf("[*] DH: Shared secret computed\n");

    /* Step 4: Derive AES key */
    dh_derive_aes_key(secret, aes_key_out);

    /* Print first 8 bytes of key as hex for verification */
    printf("[+] DH: AES-256 key derived — ");
    for (int i = 0; i < 8; i++) printf("%02x", aes_key_out[i]);
    printf("... (first 8 bytes shown)\n");

    memset(secret, 0, sizeof(secret));  /* Clear secret from stack */
    dh_free(&dh);
    return 0;
}

int perform_handshake_receiver(int sock, uint8_t aes_key_out[AES_KEY_BYTES]) {
    dh_ctx_t dh;
    if (dh_init(&dh) != 0) return -1;

    uint8_t our_pub[DH_KEY_BYTES];
    uint8_t their_pub[DH_KEY_BYTES];
    uint8_t secret[DH_KEY_BYTES];

    /* Step 1: Receive sender's public key */
    uint8_t *recv_buf = NULL;
    ssize_t  recv_len = recv_bytes(sock, &recv_buf);
    if (recv_len != DH_KEY_BYTES) {
        free(recv_buf); dh_free(&dh); return -1;
    }
    memcpy(their_pub, recv_buf, DH_KEY_BYTES);
    free(recv_buf);
    printf("[*] DH: Received sender's public key\n");

    /* Step 2: Send our public key */
    dh_get_public_key(&dh, our_pub);
    printf("[*] DH: Sending public key to sender...\n");
    if (send_bytes(sock, our_pub, DH_KEY_BYTES) != 0) {
        dh_free(&dh); return -1;
    }

    /* Step 3: Compute shared secret */
    dh_compute_shared_secret(&dh, their_pub, secret);
    printf("[*] DH: Shared secret computed\n");

    /* Step 4: Derive AES key */
    dh_derive_aes_key(secret, aes_key_out);

    printf("[+] DH: AES-256 key derived — ");
    for (int i = 0; i < 8; i++) printf("%02x", aes_key_out[i]);
    printf("... (first 8 bytes shown)\n");

    memset(secret, 0, sizeof(secret));
    dh_free(&dh);
    return 0;
}


/* ── AES-256-CBC encrypt ──────────────────────────────────────────────────── */

ssize_t aes_encrypt(const uint8_t *plaintext, size_t pt_len,
                    const uint8_t key[AES_KEY_BYTES],
                    uint8_t **out_buf) {
    /*
     * Wire format: [16-byte random IV][ciphertext with PKCS#7 padding]
     *
     * The IV is generated fresh for every call using OpenSSL's CSPRNG.
     * EVP_EncryptFinal_ex handles PKCS#7 padding automatically —
     * same padding scheme as pycryptodome in the Python implementation,
     * so C-encrypted chunks can be decrypted by the Python peer.
     */
    uint8_t iv[AES_IV_BYTES];
    if (RAND_bytes(iv, AES_IV_BYTES) != 1) return -1;

    /* Ciphertext is at most plaintext + one full padding block */
    size_t   max_ct = pt_len + AES_IV_BYTES + 16;
    uint8_t *buf    = malloc(max_ct);
    if (!buf) return -1;

    /* Prepend IV */
    memcpy(buf, iv, AES_IV_BYTES);

    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
    if (!ctx) { free(buf); return -1; }

    int out_len1 = 0, out_len2 = 0;
    EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), NULL, key, iv);
    EVP_EncryptUpdate(ctx, buf + AES_IV_BYTES, &out_len1, plaintext, (int)pt_len);
    EVP_EncryptFinal_ex(ctx, buf + AES_IV_BYTES + out_len1, &out_len2);
    EVP_CIPHER_CTX_free(ctx);

    ssize_t total = AES_IV_BYTES + out_len1 + out_len2;
    *out_buf = buf;
    return total;
}


/* ── AES-256-CBC decrypt ──────────────────────────────────────────────────── */

ssize_t aes_decrypt(const uint8_t *ciphertext, size_t ct_len,
                    const uint8_t key[AES_KEY_BYTES],
                    uint8_t **out_buf) {
    if (ct_len < AES_IV_BYTES) return -1;

    /* Split IV from ciphertext */
    const uint8_t *iv         = ciphertext;
    const uint8_t *ct         = ciphertext + AES_IV_BYTES;
    size_t         ct_data_len = ct_len - AES_IV_BYTES;

    uint8_t *buf = malloc(ct_data_len);
    if (!buf) return -1;

    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
    if (!ctx) { free(buf); return -1; }

    int out_len1 = 0, out_len2 = 0;
    EVP_DecryptInit_ex(ctx, EVP_aes_256_cbc(), NULL, key, iv);
    EVP_DecryptUpdate(ctx, buf, &out_len1, ct, (int)ct_data_len);

    if (EVP_DecryptFinal_ex(ctx, buf + out_len1, &out_len2) != 1) {
        /* Padding validation failed — data may be corrupted or tampered */
        EVP_CIPHER_CTX_free(ctx);
        free(buf);
        return -1;
    }

    EVP_CIPHER_CTX_free(ctx);
    *out_buf = buf;
    return out_len1 + out_len2;
}